### DISCUSSION TITLE
How to call the mapper for a default value other than string?

### DISCUSSION THREAD

**Discussion by DavidRomao** [mapstruct/mapstruct#3099] [Q&A] [answered]
Hello everyone,

I am looking for the best way to call the mapper method with the default value provided when the parameter is null. I've seen on the documentation this already happens if I provide a string, however if I provide an expression, which is required If I want to create an instance of another object, this expression is directly used to set the field value.

Here's a simple example in kotlin :
I have these models
```kotlin
data class Something(val config: String)
data class SomethingConfig(val id: String = "")
data class SomethingDto(val config: SomethingConfig?)
```
And this custom mapper to convert models to json
```kotlin
@Component
class JacksonJsonMapper(
    private val objectMapper: ObjectMapper
) {

    @MapToJson
    fun <R : Any> toJson(source: R?): String? {
        return source?.let { objectMapper.writeValueAsString(it) }
    }
}
```

And have defined this mapper
```kotlin
@Mapper(uses = [JacksonJsonMapper::class])
interface SomethingMapper {

    @Mapping(
        source = "config",
        target = "config",
        qualifiedBy = [MapToJson::class]
    )
    fun toEntity(source: SomethingDto): Something
}
```

Which finally generates this mapper correctly:
```java
@Component
public class SomethingMapperImpl implements SomethingMapper {

    @Autowired
    private JacksonJsonMapper jacksonJsonMapper;

    @Override
    public Something toEntity(SomethingDto source) {
        if ( source == null ) {
            return null;
        }

        String config = null;

        config = jacksonJsonMapper.toJson( source.getConfig() );

        Something something = new Something( config );

        return something;
    }
}
```

Now when source.getConfig() is null, I still want to call the method jacksonJsonMapper.toJson() with the default value SomethingConfig() so that it creates with the expected fields. The mapper would to do this for example:

```java
      if ( source.getConfig() != null ) {
            config = jacksonJsonMapper.toJson( source.getConfig() );
        }
        else {
            config = jacksonJsonMapper.toJson( new SomethingConfig() );
        }
```

I was able to achieve this by providing the default expression `jacksonJsonMapper.toJson( new SomethingConfig() )`, but I wonder if there's a better approach to achieve this ? Since this looks like a fragile solution which depends on how mapstruct names the mapper field.

Thanks in advance!

**Comment by filiphr** [comment, marked as answer]
Your current approach (using the default expression) is the only way to achieve what you are looking for right now.
